### ISSUE TITLE
Add support for all platforms

### ISSUE THREAD

**Issue by webbertakken** [webbertakken/unity-actions#12]
Keeping track of discussions and progress here, about adding support for all platforms.

- [x] Build for WebGL - v0.1
- [x] Build for Windows
- [x] Build for Linux
- [x] Build for MacOS
- [x] Build for Android
- [x] Build for iOS

The following are too specific for myself. Contributions are welcomed especially for these platforms:

- [ ] Build for Windows store
- [ ] Build for PS4
- [ ] Build for XboxOne
- [ ] Build for tvOS
- [ ] Build for Switch

The challenges with supporting multiple platforms are

- Some platforms need different input parameters
- Each platform needs a separate docker image
- The discussions about different dockerfiles and caching in https://github.com/webbertakken/unity-actions/issues/3

Please feel free to add any input anyone

**Comment by webbertakken** [comment]
This seems to be a promising solution for using different targets / images

https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#example-using-a-public-action-in-a-subdirectory

**Comment by paulpach** [comment]
Why not just let users specify the input parameters?

**Comment by webbertakken** [comment]
> Why not just let users specify the input parameters?

It's a noble idea, and I've thought about doing exactly that at first as well, but decided against it. The goal is not to make an all purpose action but rather to create independent actions with very specific tasks (following the Github Actions philosophy). That way it is easy to specialise implementations for each purpose. 

A good example is that unity builder has a default builder script included that will be run if the user doesn't provide any input. The default script will then select the enabled scenes and build them. 

This is nice because it makes builder work by default, even if the user has not created any build script yourself. This helps the end-user focus on their project, rather than technical implementation details.

Incorporating these kind of specific implementations for an all purpose action would be awkward, as it becomes unclear whether an implementing workflow is completely customised or that the user is assisted with aforementioned features.

Unless there are compelling reasons to indeed make a all-in-one action where users should specify all input parameters, I'd say this is only nice from an action-maintainer point of view and not so much for the end-user.

Anyhow; Support for multiple platforms is in and the bug that you reported has been fixed. I (or others) just have to verify each platform. That's why the issue is still open. I would appreciate any help I can get.

**Comment by webbertakken** [comment]
Closed all other issues. The only thing left for now is to start supporting all target platforms.

I do not know what is required for each platform, so I could use some input and/or help here.

**Comment by GabLeRoux** [comment]
Building for consoles is a bit tricky when it comes to doing things in public with a CI. Only Microsoft has public details as Xbox One supports UWP applications (Universal Windows Platform).

I can confirm that to this day, it is not possible to build for Xbox One using UWP with Unity running in Linux (so it's not possible to do it with gableroux/unity3d docker images). For UWP, we need Windows runners with unity installed, Visual Studio to build the generated project and Unity UWP components to accomplish this. 

That's why I made this proof of concept here:  
https://github.com/GabLeRoux/unity3d-encrypted-appveyor-poc

There is also a Xbox One component available (which is more similar to other platforms) but harder to access.

Here's some public related documentation:
* https://support.unity3d.com/hc/en-us/articles/115005327866-How-do-I-develop-to-Xbox-platforms-
* https://support.unity3d.com/hc/en-us/articles/212064606-How-do-I-develop-to-Sony-platforms-
* https://forum.unity.com/threads/how-to-make-a-nintendo-switch-game-with-unity.587791/

1. As mentioned in Unity's documentation and forums, developers need to use platform specific portals to download required components
1. Usually, developers need to sign NDAs (Non-Disclosure Agreements), request access to specific things (sdk, components) and of course, they must have access to respective developer portals.

So for other platforms, it's a bit more complicated. I think it would be more appropriate to use platform specific forums to publish CI configurations so only people with authorized access can use it. I guess there could be something like "Visit each platform developer forums and search for 'continuous integration'" or something like that.

**Comment by webbertakken** [comment]
This might also be wanted for supported platforms: https://github.com/webbertakken/unity-builder/issues/35

**Comment by davidmfinol** [comment]
I've got iOS, mac, windows, and web builds working for my project, but what I really want is Android builds. I think I'm close: https://github.com/finol-digital/Card-Game-Simulator/actions/runs/63425766/workflow

The issue I'm facing is that the Linux Unity Editor doesn't bundle the Android SDK and NDK. I'm guessing we need to download and point to them in the docker image. I can do it on my local Linux machine (and [I successfully tested a build script on my local](https://github.com/finol-digital/Card-Game-Simulator/blob/develop/Assets/Editor/BuildCGS.cs)), but I don't know how it would be done in Docker. 

Thoughts?

**Comment by GabLeRoux** [comment]
Android is blocked by the fact that images are not published with NDK yet. I plan to handle this next weekend. Details here:  
* https://gitlab.com/gableroux/unity3d/-/merge_requests/87
* https://gitlab.com/gableroux/unity3d/-/merge_requests/91
* https://gitlab.com/gableroux/unity3d/issues/61

**Comment by Rabadash8820** [comment]
@GabLeRoux Back at it again with the impressive responsiveness 😛

Android support is like the last major thing that Unity Cloud Build has over this Action (and your Docker image). Would love to see this implemented soon. Keep up the great work!

**Comment by GabLeRoux** [comment]
Just published images with NDK inside. Have a look here for versions that has support for it:  
https://gitlab.com/gableroux/unity3d/-/merge_requests/91

**Comment by davidmfinol** [comment]
[I updated my project to use Unity 2018.4.20f1](https://github.com/finol-digital/Card-Game-Simulator/actions/runs/67069703/workflow), which I believe should have the Android NDK, but I'm still getting the "Unable to locate Android NDK" error. 

Do I need to add something like EditorPrefs.SetString("AndroidNdkRoot", $"{ANDROID_NDK_PATH}") to my build script? What would the Android NDK path be?

**Comment by GabLeRoux** [comment]
@davidmfinol I've published many new images for `-android` and NDK support, but I did not target 2018 images. See description of these merge requests for the list of all images that were updated:  

* https://gitlab.com/gableroux/unity3d/-/merge_requests/100
* https://gitlab.com/gableroux/unity3d/-/merge_requests/101

I guess it's probably worth publishing new images for 2018 as well. I'll do a quick fix for `2018.4.20f1` and link the pipeline here. It will start working once the pipeline is green.

**Comment by GabLeRoux** [comment]
Nevermind, `2018.4.20f1` was part of the update, here's its build log:  
https://gitlab.com/gableroux/unity3d/-/jobs/490265858

Everything looks fine for the docker image.

I did not try `unity-actions` so for, but you have to make sure you are using the right docker image: **`gableroux/2018.4.20f1-android`**

**Comment by GabLeRoux** [comment]
Maybe 2018 requires NDK `16b`. This is the only reference I found in Unity documentation:  
https://docs.unity3d.com/2018.4/Documentation/Manual/android-sdksetup.html

**Comment by GabLeRoux** [comment]
@davidmfinol you may give a try with the following temporary image when this pipeline is green:  
https://gitlab.com/gableroux/unity3d/-/jobs/491905146

image to try: `gableroux/unity3d:2018.4.20f1-android-test-ndk-2018-16b`

**Comment by GabLeRoux** [comment]
Someone else confirmed versions before 2019.3.0 requires r16 so I did this:
https://gitlab.com/gableroux/unity3d/-/issues/70 (a bit insane, but it takes what it takes) :)

**Comment by davidmfinol** [comment]
Progress: I am no longer getting the "Unable to locate the Android NDK" error, but now, the build job is [running out of space after running for over an hour](https://github.com/finol-digital/Card-Game-Simulator/runs/549232647?check_suite_focus=true#step:5:6340).

I don't know what to make of this error. I doubt that that amount of space on the build server is actually the problem. When I run the build on my old laptop, it takes about 5 minutes, so the fact that the job runs for over an hour is a red flag. 

Thoughts on this new issue?

**Comment by webbertakken** [comment]
> Progress: I am no longer getting the "Unable to locate the Android NDK" error, but now, the build job is [running out of space after running for over an hour](https://github.com/finol-digital/Card-Game-Simulator/runs/549232647?check_suite_focus=true#step:5:6340).
> 
> I don't know what to make of this error. I doubt that that amount of space on the build server is actually the problem. When I run the build on my old laptop, it takes about 5 minutes, so the fact that the job runs for over an hour is a red flag.
> 
> Thoughts on this new issue?

To me it looks like one of the steps takes 51 minutes as indicated by [Android PostProcess task "IL2CPP" took 3064730.3235 ms](https://github.com/finol-digital/Card-Game-Simulator/runs/549232647?check_suite_focus=true#step:5:6108), and after an hour github actions pulls the filesystem from under the running action, would by my guess. Even though the [docs](https://help.github.com/en/actions/getting-started-with-github-actions/about-github-actions#usage-limits) say that execution can last up to 6 hours. 

Could you create a separate thread for this issue?

**Comment by webbertakken** [comment]
As we now have android support too, I will be closing this issue. 

Any specific platforms like xbox or playstation etc. will depend on external contributions. 

Note that they also require extra SDKs from those producers and it might become complicated because of NDAs on those SDKs.
